fem/pbilinearform.cpp

```diff
@@ -295,6 +295,50 @@ const
    pfes->Dof_TrueDof_Matrix()->MultTranspose(a, Y, 1.0, y);
 }
 
+double ParBilinearForm::InnerProduct(const ParGridFunction &x,
+                                     const ParGridFunction &y) const
+{
+   double loc = BilinearForm::InnerProduct(x, y);
+   double glob = 0.;
+
+   MPI_Allreduce(&loc, &glob, 1, MPI_DOUBLE, MPI_SUM, pfes->GetComm());
+
+   return glob;
+}
+
+double ParBilinearForm::TrueInnerProduct(const ParGridFunction &x,
+                                         const ParGridFunction &y) const
+{
+   MFEM_ASSERT(x.ParFESpace() == pfes, "the parallel spaces must match");
+   MFEM_ASSERT(y.ParFESpace() == pfes, "the parallel spaces must match");
+
+   HypreParVector *x_p = x.ParallelProject();
+   HypreParVector *y_p = y.ParallelProject();
+
+   double res = TrueInnerProduct(*x_p, *y_p);
+
+   delete x_p;
+   delete y_p;
+
+   return res;
+}
+
+double ParBilinearForm::TrueInnerProduct(HypreParVector &x,
+                                         HypreParVector &y) const
+{
+   MFEM_VERIFY(p_mat.Ptr() != NULL, "parallel matrix must be assembled");
+
+   HypreParVector *Ax = new HypreParVector(pfes);
+   HypreParMatrix *A = p_mat.As<HypreParMatrix>();
+   A->Mult(x, *Ax);
```

Comment:
This doesn't really need a  casting to a `HypreParMatrix`. It can simply be `p_mat->Mult(x,*Ax)`

Reply:
It does not, but if you look into `HypreParMatrix::Mult(const Vector&, Vector &)`, it constructs `HypreParVector`s by wrapping the T-vectors, so passing `HypreParVector`s directly saves unnecessary overhead.

Reply:
I was refering to the line  `HypreParMatrix *A = p_mat.As<HypreParMatrix>();`

Reply:
I too :wink: , but `p_mat` is not a matrix (confusingly), but `OperatorHandle`, so the multiplication would call the general routine `Mult(const Vector&, Vector&)` instead of the specific `Mult(const HypreParVector&, HypreParVector &)`.